### PR TITLE
Starting point code for MDR implementation

### PR DESCRIPTION
**Pull request by TuanNguyen27**
Assumption(s): Labels are only binary

Implemented:

fit(self, features, classes): simply build a dictionary that maps each
instance of the feature vector to a tuple. The tuple keeps count of how
many times a particular label value appears with that instance of
feature vector. Key: tuple of feature values - Value: tuple of label
frequency/label counts

transform(self, features): After the dictionary is completed, combine
each instance of feature vector above into one corresponding label that
has the frequency ratio greater than its standard default ratio.

score(self, features, classes): Compare the new combined feature vector
with its corresponding class labels, and count the times the two match.
Output the average accuracy by averaging the match count over the
length of the new feature vector / classes vector.

Implementation is tested in main() by training MDR on the training set
and getting accuracy_score on the test set.

### CODE REVIEW

File: mdr/mdr.py [hunk 1]
```diff
@@ -18,29 +18,32 @@
...
 
         Parameters
         ----------
-        param1: type (default: 1)
-            description
-        param2: type (default: True)
-            description
+        tie_break: type int (default: 0)
+            description: specify the default label in case there's a tie in a given set of feature values 
+        default_label: type int (default: 0)
```
Comment (rhiever) [inline mdr/mdr.py]: Remove the words "type"

Comment (TuanNguyen27) [inline mdr/mdr.py]: oops got cha!

File: mdr/mdr.py [hunk 2]
```diff
@@ -18,29 +18,32 @@
...
+            description: specify the default label in case there's no data for a given set of feature values  
 
         Returns
         -------
         None
 
         """
+        self.tie_break = tie_break
+        self.default_label = default_label
+        self.fdict = {}
```
Comment (rhiever) [inline mdr/mdr.py]: Please give `fdict` a more descriptive name, e.g., `feature_map`.

File: mdr/mdr.py [hunk 3]
```diff
@@ -57,6 +60,22 @@ def fit(self, features, classes):
         None
 
         """
+        n = (np.unique(classes)).size # general case 
+        for i in range(features.shape[0]):
```
Comment (rhiever) [inline mdr/mdr.py]: Give a more descriptive name than `i`.

File: mdr/mdr.py [hunk 4]
```diff
@@ -57,6 +60,22 @@ def fit(self, features, classes):
...
+            feature_instance = tuple(map(tuple, features[i])) 
+            if feature_instance not in dict: 
+                feature_map[feature_instance] = np.zeros((n,), dtype=np.int) #initialize count for a new set of feature values
```
Comment (rhiever) [inline mdr/mdr.py]: Use a `defaultdict` to avoid having to check whether a feature instance is in the dictionary.

File: mdr/mdr.py [hunk 5]
```diff
@@ -57,6 +60,22 @@ def fit(self, features, classes):
...
+        n = (np.unique(classes)).size # general case 
+        for i in range(features.shape[0]):
+            feature_instance = tuple(map(tuple, features[i])) 
+            if feature_instance not in dict: 
+                feature_map[feature_instance] = np.zeros((n,), dtype=np.int) #initialize count for a new set of feature values 
+            #assuming there are only 2 unique label values - might have to change to generalize
+            if classes[i] = 0:
+                feature_map[feature_instance][0] += 1
+            else:
+                feature_map[feature_instance][1] += 1
```
Comment (rhiever) [inline mdr/mdr.py]: This can be replaced with one line: `feature_map[feature_instance][classes[i]] += 1`

File: mdr/mdr.py [hunk 6]
```diff
@@ -73,7 +92,21 @@ def transform(self, features):
             Constructed features from the provided feature matrix
 
         """
-        return 0
+        new_feature = np.zeros(shape=(features.shape[0],1))
+        for i in range(features.shape[0]):
+            feature_instance = tuple(map(tuple, features[i]))
+            if feature_instance in feature_map:
+                counts = feature_map[feature_instance]
+                if counts[0] > counts[1]:
```
Comment (rhiever) [inline mdr/mdr.py]: This comparison needs to be made based on ratios, not counts. You calculate the overall ratio of controls (`0`) to cases (`1`) during `fit`, and store that ratio in an internal variable (`self.class_ratio`). Then the comparison will look something like:

```
if counts[0] / counts[1] > self.class_ratio:
    # ... etc.
```

File: mdr/mdr.py (hunk 5 above)
Comment (TuanNguyen27) [inline mdr/mdr.py]: got cha! one thing im not clear, if the feature_instance is not in the dict, would defaultdict know how to intialize a numpy array for it

File: mdr/mdr.py [hunk 7]
```diff
@@ -110,7 +143,11 @@ def score(self, features, classes):
             The estimated accuracy based on the constructed feature
 
         """
-        return 0.0
+        new_feature = self.fit_transform(features, classes)
```
Comment (rhiever) [inline mdr/mdr.py]: This line should only `transform`, not `fit`. The MDR model should already be fitted at this point.

That said, we should have a check here to make sure that `fit` has been called already and raise a `ValueError` if not.

File: mdr/mdr.py [hunk 8]
```diff
@@ -110,7 +143,11 @@ def score(self, features, classes):
...
+        results = (new_feature == classes).all()
```
Comment (rhiever) [inline mdr/mdr.py]: I believe `.all()` reduces the vector down to a single value. We don't want that here, as we're summing the vector in the next line.

File: mdr/mdr.py [hunk 9]
```diff
@@ -110,7 +143,11 @@ def score(self, features, classes):
...
+        score = np.sum(results)
+        accuracy_score = score/classes.size
```
Comment (rhiever) [inline mdr/mdr.py]: In general, make sure that all divisions in this code involve `float` values. Integer division will not give us the correct value.

File: mdr/mdr.py [hunk 10]
```diff
@@ -162,7 +199,9 @@ def main():
     testing_classes = input_data.loc[testing_indices, 'class'].values
 
     # Run and evaluate MDR on the training and testing data
-
+    mdr = MDR()
+    mdr.fit(training_features, training_classes)
+    return mdr.score(testing_features, testing_classes)
```
Comment (rhiever) [inline mdr/mdr.py]: For now, maybe print the score instead of returning it.

File: mdr/mdr.py (hunk 5 above)
Comment (rhiever) [inline mdr/mdr.py]: Yes. You can even do something like:

```
from collections import defaultdict
import numpy as np

classes = [0, 1, 1, 0, 1, 0] # I just made up this array
num_classes = (np.unique(classes)).size
test_dict = defaultdict(lambda: np.zeros((num_classes,), dtype=np.int))
print(test_dict['a'])
```

File: mdr/mdr.py [hunk 11]
```diff
@@ -18,29 +18,33 @@
 """
 
 import pandas as pd
-
+import numpy as np 
+from collections import defaultdict
 from __future__ import print_function
```
Comment (rhiever) [inline mdr/mdr.py]: `from __future__ import print_function` must be the first import in the file.

File: mdr/mdr.py [hunk 12]
```diff
@@ -18,29 +18,33 @@
...
+            description: specify the default label in case there's no data for a given set of feature values  
 
         Returns
         -------
         None
 
         """
+        self.tie_break = tie_break
+        self.default_label = default_label
+        self.class_ratio = 0
```
Comment (rhiever) [inline mdr/mdr.py]: Make this a float instead of an int

File: mdr/mdr.py [hunk 13]
```diff
@@ -57,6 +61,19 @@ def fit(self, features, classes):
         None
 
         """
+        self.class_ratio = float(sum(classes == 0))/(classes.size) #only applies to binary classification 
+        num_classes = (np.unique(classes)).size # count all the unique values of classes
```
Comment (rhiever) [inline mdr/mdr.py]: Raise a ValueError here if `num_classes` is not 2.

File: mdr/mdr.py [hunk 14]
```diff
@@ -57,6 +61,19 @@ def fit(self, features, classes):
         None
 
         """
+        self.class_ratio = float(sum(classes == 0))/(classes.size) #only applies to binary classification
```
Comment (rhiever) [inline mdr/mdr.py]: Rename `class_ratio` to `class_fraction`.

Use `sorted(np.unique(classes))[0]` in place of the hard-coded `0`.

File: mdr/mdr.py [hunk 15]
```diff
@@ -57,6 +61,19 @@ def fit(self, features, classes):
         None
 
         """
+        self.class_ratio = float(sum(classes == 0))/(classes.size) #only applies to binary classification 
+        num_classes = (np.unique(classes)).size # count all the unique values of classes 
+        self.feature_map = defaultdict(lambda: np.zeros((num_classes,), dtype=np.int))
+
+        for row_i in range(features.shape[0]):
+            feature_instance = tuple(map(tuple, features[row_i])) #convert feature vector to tuple
```
Comment (rhiever) [inline mdr/mdr.py]: You can probably replace this with `tuple(features[row_i])`.

File: mdr/mdr.py [hunk 16]
```diff
@@ -57,6 +61,19 @@ def fit(self, features, classes):
...
+        self.class_ratio = float(sum(classes == 0))/(classes.size) #only applies to binary classification 
+        num_classes = (np.unique(classes)).size # count all the unique values of classes 
+        self.feature_map = defaultdict(lambda: np.zeros((num_classes,), dtype=np.int))
+
+        for row_i in range(features.shape[0]):
+            feature_instance = tuple(map(tuple, features[row_i])) #convert feature vector to tuple 
+            feature_map[feature_instance][classes[row_i]] += 1 #update count 
+
+        
+
```
Comment (rhiever) [inline mdr/mdr.py:87]: `feature_map` should not be the counts. Use a variable called `class_count_matrix` instead.

Once you're done counting in `fit`, go through each tuple and calculate the class fraction for each tuple:

1) If both classes for that tuple are 0, then use `self.default_label`
2) If the class fraction == `self.class_ratio`, then use `self.tie_break`
3) If the class fraction is > `self.class_ratio`, then it is the first class
4) If the class fraction is < `self.class_ratio`, then it is the second class

Store these tuple --> class mappings in `feature_map`.

Make sure that `feature_map` is assigned as `defaultdict(lambda: self.default_label)`

File: mdr/mdr.py [hunk 17]
```diff
@@ -73,7 +90,31 @@ def transform(self, features):
             Constructed features from the provided feature matrix
 
         """
-        return 0
+        new_feature = np.zeros(shape=(features.shape[0],1), dtype=np.int)
```
Comment (rhiever) [inline mdr/mdr.py:107]: Transform should be much shorter due to the changes above. It should only go through each record in `features` and look up the tuple in `self.feature_map`. The array that is returns should be of the same length as the `features` array.

Comment (TuanNguyen27) [inline mdr/mdr.py:107]: current status: fixed all comments above, working on unit test, and will continue to commit after some more unit tests.

Comment (rhiever) [inline mdr/mdr.py:107]: Can you please commit the fixes to this branch/PR? I don't see the changes.

Comment (TuanNguyen27) [inline mdr/mdr.py:107]: I try committing again and I think all of them are there! The last one does not mark as an X but I made sure I cut down on that function too, it is just that line that was not changed.

Comment (rhiever) [inline mdr/mdr.py:107]: Maybe try syncing? Still not showing up.
